### PR TITLE
update settings schema to add 'config' section 

### PR DESCRIPTION
**Pull request by shahzebsiddiqui**
this PR is in reference to https://github.com/buildtesters/buildtest/pull/305

### CODE REVIEW

**Comment by shahzebsiddiqui** [comment]
This is ready for review.

File: settings/settings.schema.json [hunk 1]
```diff
@@ -21,11 +21,31 @@
...
       },
-      "additionalProperties": false       
+      "config": {
+        "type": "object",
+        "additionalProperties": false,
+        "properties": {
+          "paths": {
+            "type": "object",
+            "properties": {
+              "get_repo_path": {
```
Comment (vsoch) [inline settings/settings.schema.json:31]: `get_repo_path` and  `buildspec_path`? Those seem awkward.

Comment (shahzebsiddiqui) [reply settings/settings.schema.json:31]: What about repo_path or clone_path or 
For buildspec_path what about search_path or searchpath.

Any suggestions?

**Comment by shahzebsiddiqui** [comment]
> I don't fully understand the additions to the config, please walk me through these variables meaning 1) what they are, and 2) why they need to be defined in a global config.

`config` is section where buildtest configuration could go each subsection can be used to group like configuration such as `path` where we store location for directory path to where files are written such as testdir, location where buildtest clones the repos and search path for buildspec. The purpose is to help define configuration once in a file that suits everyone needs. Typically we will want to tweak location to filesystem of choice depending on amount of data generated by tests, output and any ancillary files or input data.

The config can have more section it's not limited to path. The nice thing with JSON schema is we can always change schema as new sections are added. I purposely didn't add all the sections yet but something to brainstorm on the layout.

I forsee global settings that could be variables, schedulers, logging, modules, test, debug options.
Some section can be overriden in executors and then in buildspec. The idea is to have configuration overlay so that configuration can be inherited globally, within a executor or executor types or a buildspec.

File: settings/settings.schema.json (hunk 1 above)
Comment (vsoch) [reply settings/settings.schema.json:31]: If it's nested under paths, then saying path again is redundant. If you are sure about having these sections, and you really want path, I'd just do:

```
path
    clonepath
    searchpath
```
or possibly:

```
path
    clones
    search
```

**Comment by vsoch** [comment]
Instead of hard coding specific executor variables, I would use an approach that simply updates some executor settings with a mirror of what is provided in the config, and so the validation would happen when the executor is validated, and the config/settings validation would only care about the top level (executor types). Otherwise you'll have to update this settings config every time there is a detail change to any executor, and that's likely to lead to mistakes or bugs.

**Comment by shahzebsiddiqui** [comment]
I know what you mean, you want a json definition that could be used in config and executor section. I have that in mind, but right now i don't think all config section will be applicable in the executor. Certainly not `path` subsection. I will take this into consideration as we optimize the json schema. For example `env` section in `script` could belong as a definition and be reused in other schemas and i think we need one location to put the global definitions.

File: settings/settings.schema.json (hunk 1 above)
Comment (shahzebsiddiqui) [reply settings/settings.schema.json:31]: maybe i will move it up one level what about 

```
config
  clonepath: 
  searchpath:
```

Comment (vsoch) [reply settings/settings.schema.json:31]: I think "paths" should still be its own section.

Comment (shahzebsiddiqui) [reply settings/settings.schema.json:31]: okay i took your first suggestion

**Pull request merged**